src/pyggdrasil/tree_inference/_mcmc.py

```diff
@@ -124,15 +138,23 @@ def _swap_node_labels_proposal(
     return _swap_node_labels_move(tree=tree, node1=node1, node2=node2), 0.0
 
 
-def _swap_subtrees_move(tree: Tree, node1: int, node2: int) -> Tree:
-    """Swaps subtrees rooted at ``node1`` and ``node2``.
+def _swap_subtrees_move(
+    tree: Tree,
+    node1: int,
+    node2: int,
+    same_lineage: bool,
+    key: random.PRNGKeyArray = None,
+) -> Tree:
+    """Swaps subtrees rooted at ``node1``/ ``node i`` and ``node2``/ ``node k``.
```

Comment:
This answer is going to be very subjective and opinionated, but:
  - in Python overloading is very limited, although sometimes useful for type checking – see e.g., [this tutorial](https://adamj.eu/tech/2021/05/29/python-type-hints-how-to-use-overload/);
  - whenever different sets of arguments are passed, it is a potential sign that *this function does two things and should be refactored into two functions*. (Recall [SRP](https://en.wikipedia.org/wiki/Single-responsibility_principle)).

I think this is a good example: if the case of being in the same lineage is so important, it should perfectly be in a different function. It will simplify the logic.

Then, the move (which has access to the random key) can decide which of these functions should be used (by comparing the nodes and observing whether they are in the same lineage).

I'm not sure if this writing is clear and coherent enough – let me know if not, we can also discuss it in the meeting :slightly_smiling_face:

Reply:
Very interesting! As always, I greatly apparate these nuggets of new input :) 

Tried to comply with SRP now. Split the function up into 3. 
1) `detrach the subtrees` 2) same lineage 3) different lineage

But still need to keep the choice of same or different lineage in the `swap_subtrees_proposal`, needed for the correction factor.